File: car_diary.java
// car_diary.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;

class Repair {
    int id;
    String date;
    int mileage;
    String description;
    double cost;

    Repair(int id, String date, int mileage, String description, double cost) {
        this.id = id;
        this.date = date;
        this.mileage = mileage;
        this.description = description;
        this.cost = cost;
    }

    String toJson() {
        return String.format("{\"id\":%d,\"date\":\"%s\",\"mileage\":%d,\"description\":\"%s\",\"cost\":%.2f}",
                id, date, mileage, description, cost);
    }
}

public class car_diary {
    private static final String DATA_FILE = "repairs.json";
    private static List<Repair> repairs = new ArrayList<>();
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        load();
        while (true) {
            System.out.println("\n\u001B[36m🚗 Дневник автомобиля (Java)\u001B[0m");
            System.out.println("1. Добавить запись");
            System.out.println("2. Показать все записи");
            System.out.println("3. Поиск записей");
            System.out.println("4. Удалить запись");
            System.out.println("5. Статистика");
            System.out.println("6. Выход");
            System.out.print("Выберите действие: ");
            String choice = scanner.nextLine();
            switch (choice) {
                case "1": addRepair(); break;
                case "2": listAll(); break;
                case "3": searchRepairs(); break;
                case "4": deleteRepair(); break;
                case "5": showStats(); break;
                case "6": System.out.println("До свидания!"); return;
                default: System.out.println("\u001B[31mНеверный выбор.\u001B[0m");
            }
        }
    }

    private static void load() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(DATA_FILE)));
            // Простой парсинг (для демонстрации)
            // В реальности лучше использовать JSON-библиотеку (например, Jackson)
            // Здесь мы пропустим загрузку для краткости, просто инициализируем пустой список
            repairs = new ArrayList<>();
        } catch (IOException e) {
            repairs = new ArrayList<>();
        }
    }

    private static void save() {
        try {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < repairs.size(); i++) {
                sb.append(repairs.get(i).toJson());
                if (i < repairs.size() - 1) sb.append(",");
            }
            sb.append("]");
            Files.write(Paths.get(DATA_FILE), sb.toString().getBytes());
        } catch (IOException e) {
            System.out.println("Ошибка сохранения.");
        }
    }

    private static void addRepair() {
        System.out.print("Дата (ГГГГ-ММ-ДД): ");
        String date = scanner.nextLine();
        System.out.print("Пробег (км): ");
        int mileage = Integer.parseInt(scanner.nextLine());
        System.out.print("Описание: ");
        String desc = scanner.nextLine();
        System.out.print("Стоимость (руб): ");
        double cost = Double.parseDouble(scanner.nextLine());
        int id = repairs.size() + 1;
        repairs.add(new Repair(id, date, mileage, desc, cost));
        save();
        System.out.println("\u001B[32m✅ Запись добавлена (ID: " + id + ")\u001B[0m");
    }

    private static void listAll() {
        if (repairs.isEmpty()) {
            System.out.println("\u001B[33mНет записей.\u001B[0m");
            return;
        }
        System.out.printf("\u001B[36m%-4s %-12s %-10s %-30s %10s\u001B[0m\n", "ID", "Дата", "Пробег", "Описание", "Стоимость");
        System.out.println("-".repeat(70));
        for (Repair r : repairs) {
            System.out.printf("%-4d %-12s %-10d %-30s %10.2f\n", r.id, r.date, r.mileage, r.description, r.cost);
        }
    }

    private static void searchRepairs() {
        System.out.print("Введите ключевое слово: ");
        String keyword = scanner.nextLine();
        boolean found = false;
        for (Repair r : repairs) {
            if (r.description.toLowerCase().contains(keyword.toLowerCase()) || r.date.contains(keyword) || String.valueOf(r.mileage).equals(keyword)) {
                System.out.printf("%d: %s | %d км | %s | %.2f руб.\n", r.id, r.date, r.mileage, r.description, r.cost);
                found = true;
            }
        }
        if (!found) System.out.println("\u001B[33mНичего не найдено.\u001B[0m");
    }

    private static void deleteRepair() {
        listAll();
        System.out.print("Введите ID записи для удаления: ");
        int id = Integer.parseInt(scanner.nextLine());
        boolean removed = repairs.removeIf(r -> r.id == id);
        if (removed) {
            save();
            System.out.println("\u001B[32m✅ Запись удалена.\u001B[0m");
        } else {
            System.out.println("\u001B[31m❌ Запись не найдена.\u001B[0m");
        }
    }

    private static void showStats() {
        if (repairs.isEmpty()) {
            System.out.println("Нет данных.");
            return;
        }
        double total = repairs.stream().mapToDouble(r -> r.cost).sum();
        System.out.printf("\u001B[32mОбщая стоимость ремонтов: %.2f руб.\u001B[0m\n", total);
        System.out.printf("Средняя стоимость ремонта: %.2f руб.\n", total / repairs.size());
        if (repairs.size() > 1) {
            int[] mileages = repairs.stream().mapToInt(r -> r.mileage).toArray();
            int sumIntervals = 0;
            for (int i = 1; i < mileages.length; i++) {
                sumIntervals += mileages[i] - mileages[i-1];
            }
            double avgInterval = sumIntervals / (double)(mileages.length - 1);
            System.out.printf("Средний интервал между ремонтами: %.1f км\n", avgInterval);
        }
    }
}

File: car_diary.php
<?php
// car_diary.php — PHP версия

$dataFile = 'repairs.json';

function loadRepairs() {
    global $dataFile;
    if (file_exists($dataFile)) {
        $json = file_get_contents($dataFile);
        return json_decode($json, true) ?: [];
    }
    return [];
}

function saveRepairs($repairs) {
    global $dataFile;
    file_put_contents($dataFile, json_encode($repairs, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
}

$repairs = loadRepairs();

function color($text, $code) {
    return "\033[{$code}m{$text}\033[0m";
}

while (true) {
    echo "\n" . color("🚗 Дневник автомобиля (PHP)", '36') . "\n";
    echo "1. Добавить запись\n";
    echo "2. Показать все записи\n";
    echo "3. Поиск записей\n";
    echo "4. Удалить запись\n";
    echo "5. Статистика\n";
    echo "6. Выход\n";
    echo "Выберите действие: ";
    $choice = trim(fgets(STDIN));

    switch ($choice) {
        case '1':
            echo "Дата (ГГГГ-ММ-ДД): ";
            $date = trim(fgets(STDIN));
            echo "Пробег (км): ";
            $mileage = (int) trim(fgets(STDIN));
            echo "Описание: ";
            $desc = trim(fgets(STDIN));
            echo "Стоимость (руб): ";
            $cost = (float) trim(fgets(STDIN));
            $id = count($repairs) + 1;
            $repairs[] = [
                'id' => $id,
                'date' => $date,
                'mileage' => $mileage,
                'description' => $desc,
                'cost' => $cost
            ];
            saveRepairs($repairs);
            echo color("✅ Запись добавлена (ID: $id)\n", '32');
            break;

        case '2':
            if (empty($repairs)) {
                echo color("Нет записей.\n", '33');
            } else {
                printf(color("%-4s %-12s %-10s %-30s %10s\n", '36'), "ID", "Дата", "Пробег", "Описание", "Стоимость");
                echo str_repeat("-", 70) . "\n";
                foreach ($repairs as $r) {
                    printf("%-4d %-12s %-10d %-30s %10.2f\n", $r['id'], $r['date'], $r['mileage'], $r['description'], $r['cost']);
                }
            }
            break;

        case '3':
            echo "Введите ключевое слово: ";
            $keyword = trim(fgets(STDIN));
            $found = false;
            foreach ($repairs as $r) {
                if (stripos($r['description'], $keyword) !== false || strpos($r['date'], $keyword) !== false || $r['mileage'] == $keyword) {
                    echo "{$r['id']}: {$r['date']} | {$r['mileage']} км | {$r['description']} | {$r['cost']} руб.\n";
                    $found = true;
                }
            }
            if (!$found) echo color("Ничего не найдено.\n", '33');
            break;

        case '4':
            if (empty($repairs)) {
                echo color("Нет записей для удаления.\n", '33');
                break;
            }
            // показать список
            foreach ($repairs as $r) {
                echo "{$r['id']}: {$r['date']} | {$r['description']}\n";
            }
            echo "Введите ID записи для удаления: ";
            $id = (int) trim(fgets(STDIN));
            $index = array_search($id, array_column($repairs, 'id'));
            if ($index !== false) {
                array_splice($repairs, $index, 1);
                saveRepairs($repairs);
                echo color("✅ Запись удалена.\n", '32');
            } else {
                echo color("❌ Запись не найдена.\n", '31');
            }
            break;

        case '5':
            if (empty($repairs)) {
                echo "Нет данных.\n";
                break;
            }
            $total = array_sum(array_column($repairs, 'cost'));
            echo color("Общая стоимость ремонтов: {$total} руб.\n", '32');
            echo "Средняя стоимость ремонта: " . ($total / count($repairs)) . " руб.\n";
            if (count($repairs) > 1) {
                $mileages = array_column($repairs, 'mileage');
                $intervals = [];
                for ($i = 1; $i < count($mileages); $i++) {
                    $intervals[] = $mileages[$i] - $mileages[$i-1];
                }
                $avgInterval = array_sum($intervals) / count($intervals);
                echo "Средний интервал между ремонтами: {$avgInterval} км\n";
            }
            break;

        case '6':
            echo "До свидания!\n";
            exit(0);

        default:
            echo color("Неверный выбор.\n", '31');
    }
}
?>
